File: src/data-structures/stack/stack.ts
export function stack<T>() {
  /**
   * storage will be private and not exported
   */
  const storage: T[] = []
  /**
   *
   * @param item that we push into the Stack
   * Add to the end, bottom.
   *
   */
  const push = (item: T) => {
    storage.push(item)
  }

  /**
   * Removers from the top
   * The first item in the Stack
   */
  const pop = () => {
    storage.pop()
  }

  const isEmpty = (): boolean => {
    return storage.length === 0
  }

  /**
   * Returns the item from the top without removing it
   */
  const peek = (): T | T[] => {
    return storage.length > 0 ? storage[storage.length - 1] : storage
  }

  return {
    push,
    pop,
    peek,
    isEmpty,
  }
}

// const myStack = stack()

// myStack.push(45)
// myStack.push(25)
// myStack.push(15)

// myStack.pop()
// myStack.pop()
// myStack.pop()
// console.log(myStack.isEmpty())

// console.log(myStack.peek())

File: src/data-structures/stack/stack.go
package stack

// Storage struct
type Storage []string

// IsEmpty check if is empty
func (s *Storage) IsEmpty() bool {
	return len(*s) == 0
}

// Push on the top of the stack
func (s *Storage) Push(item string) {
	*s = append(*s, item)
}

// Pop removes the first item from the top
func (s *Storage) Pop() (string, bool) {
	if s.IsEmpty() {
		return "empty", false
	} else {
		index := len(*s) - 1
		element := (*s)[index]
		*s = (*s)[:index]
		return element, true
	}
}

// Peek return the top element
func (s *Storage) Peek() (string, bool) {
	if s.IsEmpty() {
		return "empty", false
	} else {
		return (*s)[len(*s)-1], true
	}
}

// Stack fn
func Stack() Storage {

	var storage Storage
	storage.Push("apple")
	storage.Push("banana")
	storage.Push("resberry")
	storage.Pop()
	

	return storage
}
